Revive remove default ghosting

lindsayad opened the pull request:
This eliminates libMesh's default ghosting for MOOSE simulations, and only adds ghosting when it is needed, e.g. when we have objects like:

- adaptivity
- DG
- interface kernels
- internal side user objects
- several output objects 
- `MultiAppProjectionTransfer`

and more in modules.

friedmud committed:
Add GhostingAux to display ghosted elements - might not be working quite right yet refs #12298

friedmud committed:
Use new interfaces in libMesh for disabling default ghosting refs #12327

friedmud committed:
Add tests and doco for GhostingAux refs #12298 #12327

friedmud committed:
Update libmesh to get the new APIs for disabling default coupling refs #12327

friedmud committed:
Work in progress towards #12354

friedmud committed:
More work towards #12354

friedmud committed:
Big refactor of RelationshipManger refs #12354

friedmud committed:
Fixup issue in SetAdaptivityOptionsAction refs #12354

friedmud committed:
Make stringify understand geometric and algebraic refs #12354

friedmud committed:
All adaptive tests pass refs #12354

friedmud committed:
DGKernels working with relationship manager #12354

friedmud committed:
All tests passing refs #12354

friedmud committed:
Fixup tests when running with distributed mesh refs #12354

friedmud committed:
Fixup phase_field refs #12354

friedmud committed:
Fixup compile error refs #12354

friedmud committed:
Fixup another compile issue refs #12354

friedmud committed:
Add missing include for non-unity build refs #12354

friedmud committed:
Regold test refs #12354

friedmud committed:
Fixup includes refs #12354

friedmud committed:
Regold refs #12354

friedmud committed:
Use grid partitioner for more stability refs #12354

friedmud committed:
Add some ghosting to FeatureFloodCount refs #12354

friedmud committed:
Attempt to add ProxyRelationshipManager, but its probably not going to work with adaptivity refs #12354

lindsayad committed:
Make ghost syncing between ref and displaced work with adaptivity

lindsayad committed:
Mimic somewhat libmesh's default ghosting when performing adaptivity

All moose tests are passing on my Mac when running with a simple run_tests command

lindsayad committed:
Make sure to do call backs for all GhostingFunctor API methods

lindsayad committed:
Fix up modules

lindsayad committed:
Correct includes for non-unity builds

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/313248/) on fcac92b wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/13145/site)

This comment will be updated on new commits.

permcody commented on framework/src/actions/AdaptivityAction.C:
```diff
@@ -83,62 +86,122 @@ AdaptivityAction::AdaptivityAction(InputParameters params) : Action(params) {}
+    const std::vector<Real> & weight_values = getParam<std::vector<Real>>("weight_values");
+
+    int num_weight_names = weight_names.size();
+    int num_weight_values = weight_values.size();
```
What's with the `int`s? Shouldn't this be `std::size_t` or `auto`?

permcody commented on framework/src/actions/AdaptivityAction.C:
```diff
@@ -83,62 +86,122 @@ AdaptivityAction::AdaptivityAction(InputParameters params) : Action(params) {}
+      for (int i = 0; i < num_weight_names; i++)
+      {
+        std::string name = weight_names[i];
+        double value = weight_values[i];
```
`double`?

permcody commented on framework/src/actions/AdaptivityAction.C:
```diff
@@ -83,62 +86,122 @@ AdaptivityAction::AdaptivityAction(InputParameters params) : Action(params) {}
+      std::vector<Real> weights(system.nVariables(), 0);
 
-    std::vector<FEMNormType> norms(system.nVariables(), H1_SEMINORM);
+      for (int i = 0; i < num_weight_names; i++)
```
`int`?

permcody commented on framework/src/actions/AddPeriodicBCAction.C:
```diff
@@ -146,66 +148,103 @@ AddPeriodicBCAction::autoTranslationBoundaries()
+    {
+      auto rm_obj =
+          _factory.create<RelationshipManager>("ElementSideNeighborLayers",
+                                               "periodc_bc_ghosting" + Moose::stringify(count++),
```
I'm not sure I love this `count` business. Why not use the `for_whom` field here instead? I don't think we need to add more than one of these per field.

permcody commented on framework/src/actions/AddRelationshipManager.C:
```diff
@@ -26,13 +26,29 @@ AddRelationshipManager::AddRelationshipManager(InputParameters params) : Action(
+    for (const auto & action_ptr : all_action_ptrs)
+      action_ptr->addRelationshipManagers(Moose::RelationshipManagerType::ALGEBRAIC);
+
+    _app.attachRelationshipManagers(Moose::RelationshipManagerType::ALGEBRAIC);
```
Funny - this manual unroll created some code duplication. I think I liked it the other way better.

permcody commented on framework/src/auxkernels/GhostingAux.C:
```diff
@@ -0,0 +1,83 @@
+}
+
+void
+GhostingAux::initialize()
```
Still don't love this, but I guess that ship has sailed.

permcody commented on framework/src/auxkernels/GhostingAux.C:
```diff
@@ -0,0 +1,83 @@
+    mooseError("GhostingAux only works on elemental fields.");
+
+  // Can only work on Replicated meshes
+  _mesh.errorIfDistributedMesh("GhostingAux");
```
This _really_ should be made to work with DistributedMesh. Figuring out the ghosting on ReplicatedMesh feels more like a testing exercise. Is this a MOOSE utility or not? This object is adding a new API to AuxKernel so it really should be first class IMHO.

permcody commented on framework/src/base/Moose.C:
```diff
@@ -201,6 +203,8 @@ addActionTypes(Syntax & syntax)
   registerTask("setup_function_complete", false);
   registerTask("setup_variable_complete", false);
   registerTask("ready_to_init", true);
+  appendMooseObjectTask("ready_to_init", InterfaceKernel);
```
I'd prefer if these lines were moved up and clustered with existing `ready_to_init` tasks in the section up above.

lindsayad replied:
Huh? The only three occurrences of `ready_to_init` are right here.

permcody replied:
oh weird... I'm not sure if this is a bug or not. Note the name of the method used here. It's called "appendMooseObjectTask", which is supposed to correspond with "registerMooseObjectTask". The first registration or `ready_to_init` is not a MOOSE object task so I hope that the code is doing the right thing. Note that this is definitely a unique combination of these registration functions so it should probably be checked.

permcody commented on framework/src/base/MooseApp.C:
```diff
@@ -1687,10 +1705,65 @@ MooseApp::addRelationshipManager(std::shared_ptr<RelationshipManager> relationsh
 
 void
-MooseApp::attachRelationshipManagers(Moose::RelationshipManagerType rm_type)
+MooseApp::attachRelationshipManagers(const Moose::RelationshipManagerType & rm_type)
```
This is an enum. Not sure we need to pass it by reference.

permcody commented on framework/src/relationshipmanagers/ProxyRelationshipManager.C:
```diff
@@ -0,0 +1,107 @@
+bool
+ProxyRelationshipManager::operator==(const RelationshipManager & /*rhs*/) const
+{
+  // There isn't a need to determine these because only the correct ones will be added
```
There's confidence!

permcody commented on framework/src/utils/Conversion.C:
```diff
@@ -346,25 +346,32 @@ stringToEnum<RelationshipManagerType>(const std::string & s)
-    default:
-      return "ERROR";
-  }
+  // Connot make a switch statement because the boolean logic doesn't work well with the class type
```
Cannot

permcody commented on modules/phase_field/src/auxkernels/FeatureFloodCountAux.C:
```diff
@@ -39,6 +39,10 @@ validParams<FeatureFloodCountAux>()
                              "(UNIQUE_REGION and VARIABLE_COLORING are nodal, CENTROID is "
                              "elemental, default: UNIQUE_REGION)");
 
+  params.addRelationshipManager("ElementSideNeighborLayers",
```
This object doesn't need a relationship manager. It operates on local elements only. Now the FeatureFloodCount UO *does* need an RM.

permcody commented on modules/phase_field/src/postprocessors/GrainTracker.C:
```diff
@@ -52,9 +52,18 @@ validParams<GrainTracker>()
+      "ElementPointNeighborLayers",
+      Moose::RelationshipManagerType::GEOMETRIC,
+
+      [](const InputParameters & obj_params, InputParameters & rm_params) {
```
👍

permcody reviewed (commented):
I don't see any showstoppers here. You should probably address comments and then let's plan to merge this in and fixup apps.

lindsayad committed:
Address review

moosebuild commented:
Job [Boost enabled, Pthreads only](https://civet.inl.gov/job/315534/) on a281ef9 : invalidated by @lindsayad

lindsayad committed:
More aggressive ghosting for output mesh serialization

I was experiencing occasional failures using the ElementSideNeighborLayers
ghosting functor when running output tests with say GMV or Tecplot
outputs. Presumably the occasional failures can be attributed to different
partitions from ParMetis. Pre-this-PR these tests would pass even on our
highest distributed mesh sweeps, so I feel like we should be able to make
them pass now. Using @roystgnr's default geometric ghosting functor
`GhostPointNeighbors` appears to do the trick

permcody reviewed (approved):
Approved with trepidation
